File: java/src/main/java/global/thalion/ttio/codecs/ReferenceResolver.java
/*
 * TTI-O Java Implementation
 * Copyright (c) 2026 The Thalion Initiative
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */
package global.thalion.ttio.codecs;

import global.thalion.ttio.hdf5.Hdf5File;
import global.thalion.ttio.hdf5.Hdf5Group;
import global.thalion.ttio.providers.StorageDataset;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.util.List;

/**
 * Resolve a reference chromosome sequence for REF_DIFF decode.
 *
 * <p>Lookup chain (per M93 design Q5c — hard error on miss):
 * <ol>
 *   <li>Embedded {@code /study/references/<uri>/} in the open .tio file.</li>
 *   <li>External FASTA at {@code externalReferencePath} or
 *       {@code REF_PATH} environment variable.</li>
 *   <li>{@link RefMissingException} (no partial decode).</li>
 * </ol>
 *
 * <p>The MD5 attribute on the embedded reference group is verified
 * against the {@code expectedMd5} argument; mismatches raise
 * {@link RefMissingException} rather than silently returning the wrong
 * sequence.
 *
 * <p>Cross-language equivalents:
 * <ul>
 *   <li>Python: {@code ttio.genomic.reference_resolver.ReferenceResolver}</li>
 *   <li>Objective-C: {@code TTIOReferenceResolver}</li>
 * </ul>
 */
public final class ReferenceResolver {

    private final Hdf5File h5File;
    private final Path externalReferencePath;

    /** Construct a resolver bound to an open HDF5 file. The external
     *  reference path defaults to the {@code REF_PATH} env var when
     *  {@code externalReferencePath} is {@code null}. */
    public ReferenceResolver(Hdf5File h5File, Path externalReferencePath) {
        this.h5File = h5File;
        this.externalReferencePath = externalReferencePath != null
            ? externalReferencePath : envPath();
    }

    /** Convenience constructor: external path falls back to
     *  {@code REF_PATH}. */
    public ReferenceResolver(Hdf5File h5File) {
        this(h5File, null);
    }

    private static Path envPath() {
        String p = System.getenv("REF_PATH");
        return (p != null && !p.isEmpty()) ? Path.of(p) : null;
    }

    /** Return the chromosome's reference sequence as uppercase ACGTN
     *  bytes.
     *
     *  @throws RefMissingException when the reference can't be found or
     *      its MD5 doesn't match. */
    public byte[] resolve(String uri, byte[] expectedMd5, String chromosome) {
        // 1. Try embedded.
        if (h5File != null) {
            try (Hdf5Group root = h5File.rootGroup()) {
                if (root.hasChild("study")) {
                    try (Hdf5Group study = root.openGroup("study")) {
                        if (study.hasChild("references")) {
                            try (Hdf5Group refs = study.openGroup("references")) {
                                if (refs.hasChild(uri)) {
                                    return resolveEmbedded(refs, uri,
                                        expectedMd5, chromosome);
                                }
                            }
                        }
                    }
                }
            }
        }

        // 2. Try external FASTA.
        if (externalReferencePath != null
            && Files.exists(externalReferencePath)) {
            byte[] seq = externalChromosome(externalReferencePath, expectedMd5, chromosome);
            if (seq != null) {
                return seq;
            }
        }

        // 3. Hard error.
        String envValue = System.getenv("REF_PATH");
        throw new RefMissingException(
            "reference '" + uri + "' (chromosome '" + chromosome
            + "') not found in file's /study/references/ and not resolvable "
            + "via REF_PATH (" + (envValue == null ? "<unset>" : envValue)
            + "). Provide via externalReferencePath constructor arg or set "
            + "REF_PATH.");
    }

    private static byte[] resolveEmbedded(
        Hdf5Group refs, String uri, byte[] expectedMd5, String chromosome) {
        try (Hdf5Group refGrp = refs.openGroup(uri)) {
            String md5Hex = refGrp.readStringAttribute("md5");
            byte[] embeddedMd5 = hexToBytes(md5Hex);
            if (!java.util.Arrays.equals(embeddedMd5, expectedMd5)) {
                throw new RefMissingException(
                    "MD5 mismatch for embedded reference '" + uri
                    + "': expected " + bytesToHex(expectedMd5)
                    + ", got " + md5Hex);
            }
            if (!refGrp.hasChild("chromosomes")) {
                throw new RefMissingException(
                    "embedded reference '" + uri
                    + "' has no chromosomes/ subgroup");
            }
            try (Hdf5Group chroms = refGrp.openGroup("chromosomes")) {
                if (!chroms.hasChild(chromosome)) {
                    List<String> covered = chroms.childNames();
                    java.util.Collections.sort(covered);
                    throw new RefMissingException(
                        "chromosome '" + chromosome
                        + "' not embedded in reference '" + uri
                        + "' — covered_chromosomes are " + covered);
                }
                try (Hdf5Group chrom = chroms.openGroup(chromosome)) {
                    // Wrap as StorageGroup for the layout dispatch:
                    // data_packed (2-bit + run mask) when present,
                    // legacy raw data otherwise.
                    var adapter = global.thalion.ttio.providers
                        .Hdf5Provider.adapterForGroup(chrom);
                    return global.thalion.ttio.genomics.PackedReference
                        .readChromosomeBytes(adapter);
                }
            }
        }
    }

    /** Tiny FASTA reader — extract a single chromosome's sequence as
     *  bytes. Returns {@code null} if the chromosome is not present.
     *  Matches headers on the first whitespace-delimited token after
     *  {@code >}. */
    private static final java.util.Map<Path, global.thalion.ttio.genomics.LazyReference> LAZY =
        new java.util.concurrent.ConcurrentHashMap<>();
    private static final java.util.Map<Path, byte[]> SET_MD5 =
        new java.util.concurrent.ConcurrentHashMap<>();

    /** Read {@code chromosome} from the external FASTA through its .fai
     *  index and check {@code expectedMd5} against, in order: the md5 of
     *  that chromosome's case-preserved bytes, of its upper-cased bytes
     *  (both the pre-1.9 external check, which only ever matched a
     *  single-contig FASTA), then the reference-set md5 of the whole
     *  FASTA (every chromosome, alphabetic order, case preserved: the
     *  digest the writers record). Returns the upper-cased sequence, or
     *  null when the chromosome is not in the FASTA. */
    private static byte[] externalChromosome(Path path, byte[] expectedMd5, String chromosome) {
        Path key = path.toAbsolutePath();
        global.thalion.ttio.genomics.LazyReference ref =
            LAZY.computeIfAbsent(key, k -> new global.thalion.ttio.genomics.LazyReference(k, 2));
        if (!ref.containsKey(chromosome)) {
            return null;
        }
        byte[] raw = ref.get(chromosome);
        byte[] upper = new byte[raw.length];
        for (int i = 0; i < raw.length; i++) {
            byte b = raw[i];
            upper[i] = (b >= 'a' && b <= 'z') ? (byte) (b - 32) : b;
        }
        if (java.util.Arrays.equals(md5(raw), expectedMd5)
                || java.util.Arrays.equals(md5(upper), expectedMd5)) {
            return upper;
        }
        byte[] setMd5 = SET_MD5.computeIfAbsent(key, k -> ref.setMd5());
        if (java.util.Arrays.equals(setMd5, expectedMd5)) {
            return upper;
        }
        throw new RefMissingException(
            "MD5 mismatch for external reference at " + path + ": expected "
            + bytesToHex(expectedMd5) + ", got " + bytesToHex(setMd5)
            + " for the whole FASTA and " + bytesToHex(md5(raw))
            + " for chromosome " + chromosome);
    }

    private static byte[] md5(byte[] data) {
        try {
            return MessageDigest.getInstance("MD5").digest(data);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static byte[] hexToBytes(String hex) {
        if (hex == null || (hex.length() & 1) != 0) {
            throw new RefMissingException(
                "embedded @md5 attribute is not even-length hex: " + hex);
        }
        byte[] out = new byte[hex.length() / 2];
        for (int i = 0; i < out.length; i++) {
            int hi = Character.digit(hex.charAt(2 * i), 16);
            int lo = Character.digit(hex.charAt(2 * i + 1), 16);
            if (hi < 0 || lo < 0) {
                throw new RefMissingException(
                    "embedded @md5 attribute contains non-hex chars: " + hex);
            }
            out[i] = (byte) ((hi << 4) | lo);
        }
        return out;
    }

    private static String bytesToHex(byte[] buf) {
        StringBuilder sb = new StringBuilder(buf.length * 2);
        for (byte b : buf) sb.append(String.format("%02x", b & 0xFF));
        return sb.toString();
    }
}

File: python/src/ttio/genomic/reference_resolver.py
"""Reference resolver for the M93 REF_DIFF codec.

Lookup chain (per Q5c = hard error in the M93 design spec):

    embedded /study/references/<uri>/ in the open .tio file
        → external REF_PATH env var (or explicit external_reference_path=)
        → RefMissingError (no partial decode).

The resolver yields a chromosome's full uppercase ACGTN bytes. The
encoded MD5 attribute on the embedded reference group is verified
against the ``expected_md5`` argument; mismatches raise
:class:`RefMissingError` rather than silently returning the wrong
sequence.

Cross-language: ObjC ``TTIOReferenceResolver``; Java
``codecs.ReferenceResolver``.
"""
from __future__ import annotations

import hashlib
import os
from pathlib import Path
from typing import TYPE_CHECKING

import numpy as np

if TYPE_CHECKING:  # pragma: no cover — annotation only
    from ..providers.base import StorageGroup


def _hex_str_attr(raw: object) -> str:
    """Coerce an h5py attribute (bytes / numpy scalar / str) to a hex str."""
    if isinstance(raw, bytes):
        return raw.decode("ascii")
    if isinstance(raw, np.bytes_):
        return raw.tobytes().decode("ascii")
    if isinstance(raw, str):
        return raw
    if isinstance(raw, np.ndarray) and raw.size == 1:
        item = raw.item()
        if isinstance(item, bytes):
            return item.decode("ascii")
        return str(item)
    return str(raw)


class RefMissingError(RuntimeError):
    """Raised when a reference required for REF_DIFF decode cannot be resolved.

    Per M93 design spec Q5c: hard error rather than partial decode.
    Genomic data integrity is non-negotiable.
    """


class ReferenceResolver:
    """Resolve a reference chromosome sequence for REF_DIFF decode.

    Args:
        references_group: the ``/study/references`` group as a
            :class:`~ttio.providers.base.StorageGroup`, or ``None`` when
            no embedded references are available. The resolver looks for
            ``<uri>/`` children under this group as the primary source.
        external_reference_path: optional explicit path to a FASTA file.
            If unset, the ``REF_PATH`` environment variable is consulted.
    """

    def __init__(
        self,
        references_group: "StorageGroup | None" = None,
        external_reference_path: Path | None = None,
    ):
        self._refs = references_group
        self._external = external_reference_path or self._env_path()

    @staticmethod
    def _env_path() -> Path | None:
        ref_path = os.environ.get("REF_PATH")
        return Path(ref_path) if ref_path else None

    def resolve(self, uri: str, expected_md5: bytes, chromosome: str) -> bytes:
        """Return the chromosome's reference sequence as uppercase ACGTN bytes.

        Raises:
            RefMissingError: when the reference can't be found or its
                MD5 doesn't match.
        """
        # 1. Try embedded.
        if self._refs is not None and self._refs.has_child(uri):
            ref_grp = self._refs.open_group(uri)
            embedded_md5 = bytes.fromhex(_hex_str_attr(ref_grp.get_attribute("md5")))
            if embedded_md5 != expected_md5:
                raise RefMissingError(
                    f"MD5 mismatch for embedded reference {uri!r}: "
                    f"expected {expected_md5.hex()}, got {embedded_md5.hex()}"
                )
            chroms = ref_grp.open_group("chromosomes")
            if not chroms.has_child(chromosome):
                raise RefMissingError(
                    f"chromosome {chromosome!r} not embedded in "
                    f"reference {uri!r} — covered_chromosomes are "
                    f"{sorted(chroms.child_names())}"
                )
            from . import packed_reference
            return packed_reference.read_chromosome_bytes(
                chroms.open_group(chromosome))

        # 2. Try external FASTA.
        if self._external is not None and self._external.exists():
            seq = _external_chromosome(self._external, expected_md5, chromosome)
            if seq is not None:
                return seq

        # 3. Hard error (Q5c).
        raise RefMissingError(
            f"reference {uri!r} (chromosome {chromosome!r}) not found in "
            f"file's /study/references/ and not resolvable via REF_PATH "
            f"({os.environ.get('REF_PATH', '<unset>')}). Provide via "
            f"external_reference_path= constructor arg or set REF_PATH."
        )


_LAZY: dict[Path, "object"] = {}
_SET_MD5: dict[Path, bytes] = {}


def _lazy(path: Path):
    from .lazy_reference import LazyReference
    ref = _LAZY.get(path)
    if ref is None:
        ref = _LAZY[path] = LazyReference(path, cache_chroms=2)
    return ref


def _external_chromosome(path: Path, expected_md5: bytes, chromosome: str) -> bytes | None:
    """Read ``chromosome`` from the external FASTA through its .fai index
    and check ``expected_md5`` against, in order: the md5 of that
    chromosome's case-preserved bytes, of its upper-cased bytes (both
    the pre-1.9 external check, which only ever matched a
    single-contig FASTA), then the reference-set md5 of the whole
    FASTA (every chromosome, alphabetic order, case preserved: the
    digest the writers record). Returns the upper-cased sequence, or
    None when the chromosome is not in the FASTA."""
    ref = _lazy(path)
    if chromosome not in ref:
        return None
    raw = ref[chromosome]
    upper = raw.upper()
    if hashlib.md5(raw).digest() == expected_md5 or hashlib.md5(upper).digest() == expected_md5:
        return upper
    set_md5 = _SET_MD5.get(path)
    if set_md5 is None:
        set_md5 = _SET_MD5[path] = ref.set_md5()
    if set_md5 == expected_md5:
        return upper
    raise RefMissingError(
        f"MD5 mismatch for external reference at {path}: expected "
        f"{expected_md5.hex()}, got {set_md5.hex()} for the whole FASTA and "
        f"{hashlib.md5(raw).hexdigest()} for chromosome {chromosome!r}"
    )
